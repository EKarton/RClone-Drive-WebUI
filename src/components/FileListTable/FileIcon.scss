.file-icon {
  &__preview-icon {
    display: flex;
    align-items: center;
  }

  &__img {
    width: 100%;
    height: 100%;
    padding: 0 !important;
    object-fit: contain;
  }
}
