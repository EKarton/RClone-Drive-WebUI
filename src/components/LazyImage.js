import { useInView } from 'react-intersection-observer';
import Image from './Image';
import './LazyImage.scss';

export default function LazyImage({ image, width, height, imgClassName }) {
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: '0px',
  });

  const renderContent = () => {
    if (!inView) {
      return null;
    }

    return (
      <Image
        image={image}
        width={width}
        height={height}
        imgClassName="lazy-image__img"
        skeletonClassName="lazy-image__skeleton"
      />
    );
  };

  return (
    <div ref={ref} data-inview={inView} style={{ width, height }}>
      {renderContent()}
    </div>
  );
}
