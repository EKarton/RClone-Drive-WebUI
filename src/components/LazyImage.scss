.lazy-image {
  &__img {
    object-fit: cover;
    padding: 0.125rem;
  }

  &__skeleton-wrapper {
    width: 100% !important;
    height: 100% !important;
    padding: 0.125rem;
  }

  &__skeleton {
    width: 100% !important;
    height: 100% !important;
  }
}
