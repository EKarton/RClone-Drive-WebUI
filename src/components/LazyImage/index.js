import PropTypes from 'prop-types';
import { useInView } from 'react-intersection-observer';
import Image from 'components/Image';
import './index.scss';

export default function LazyImage({ image, width, height }) {
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: '0px',
  });

  const renderContent = () => {
    if (!inView) {
      return null;
    }

    return (
      <Image
        image={image}
        width={width}
        height={height}
        imgClassName="lazy-image__img"
        skeletonClassName="lazy-image__skeleton"
      />
    );
  };

  return (
    <div ref={ref} data-inview={inView} style={{ width, height }}>
      {renderContent()}
    </div>
  );
}

LazyImage.propTypes = {
  image: Image.propTypes.image,
  width: PropTypes.number,
  height: PropTypes.number,
};
