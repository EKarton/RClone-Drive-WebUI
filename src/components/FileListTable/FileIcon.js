import FolderIcon from '@mui/icons-material/Folder';
import DescriptionIcon from '@mui/icons-material/Description';
import ImageIcon from '@mui/icons-material/Image';
import Icon from '@mui/material/Icon';
import './FileIcon.scss';
import LazyImage from 'components/LazyImage';

export default function FileIcon({ file, iconSize }) {
  if (file.isDirectory) {
    return <FolderIcon fontSize={iconSize} />;
  }

  if (file.isImage) {
    return (
      <Icon fontSize={iconSize} className="file-icon__preview-icon">
        <LazyImage
          image={{
            remote: file.remote,
            folderPath: file.path,
            fileName: file.name,
          }}
          imgClassName="file-icon__img"
        />
      </Icon>
    );
  }

  if (file.isImage) {
    return <ImageIcon fontSize={iconSize} />;
  }

  return <DescriptionIcon fontSize={iconSize} />;
}
