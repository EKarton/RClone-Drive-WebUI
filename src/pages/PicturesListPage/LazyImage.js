import { Skeleton } from "@mui/material";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import { useInView } from "react-intersection-observer";
import "./LazyImage.scss";

const LazyImage = ({ image, width, height }) => {
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: "0px",
  });

  const rCloneClient = useRCloneClient();
  const [imageUrl, setImageUrl] = useState();
  const [error, setError] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setError(undefined);
        setImageUrl(undefined);

        const imageBlob = await rCloneClient.fetchFileContents(
          image.remote,
          image.folderPath,
          image.fileName
        );

        const imageUrl = URL.createObjectURL(new Blob([imageBlob]));
        setImageUrl(imageUrl);
      } catch (error) {
        setError(error);
      }
    };

    if (inView && !imageUrl && image) {
      fetchData();
    }

    return () => {
      if (imageUrl) {
        URL.revokeObjectURL(imageUrl);
      }
    };
  }, [image, imageUrl, inView, rCloneClient]);

  const renderContent = () => {
    if (error) {
      return <div>Error! {error.message}</div>;
    }

    if (!imageUrl) {
      return (
        <Skeleton variant="rectangular" className="lazy-image__skeleton" />
      );
    }

    return (
      <img
        className="lazy-image__img"
        src={imageUrl}
        alt={image.fileName}
        loading="lazy"
        width={width}
        height={height}
      />
    );
  };

  return (
    <div ref={ref} data-inview={inView} style={{ width, height }}>
      {inView ? renderContent() : null}
    </div>
  );
};

export default LazyImage;
