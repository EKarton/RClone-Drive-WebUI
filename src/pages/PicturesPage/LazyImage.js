import { Skeleton } from "@mui/material";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import Image from "./Image";
import { useInView } from "react-intersection-observer";
import "./LazyImage.scss";

const LazyImage = ({ remote, folderPath, fileName }) => {
  const { ref, inView } = useInView({
    triggerOnce: true,
    rootMargin: "0px",
  });

  const [height, setHeight] = useState("200px");

  const rCloneClient = useRCloneClient();
  const [imageUrl, setImageUrl] = useState();
  const [error, setError] = useState();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setError(undefined);
        setImageUrl(undefined);

        const imageBlob = await rCloneClient.fetchFileContents(
          remote,
          folderPath,
          fileName
        );

        const imageUrl = window.URL.createObjectURL(new Blob([imageBlob]));
        setImageUrl(imageUrl);
      } catch (error) {
        setError(error);
      }
    };

    if (inView && !imageUrl) {
      fetchData();
    }
  }, [fileName, folderPath, imageUrl, inView, rCloneClient, remote]);

  const reAdjustContainerHeight = (imgElement) => {
    const renderedImgWidth = imgElement.clientWidth;
    const actualImgWidth = imgElement.naturalWidth;
    const actualImgHeight = imgElement.naturalHeight;

    const aspectRatio = actualImgHeight / actualImgWidth;

    setHeight(renderedImgWidth * aspectRatio);
  };

  const handleImageResized = (e) => {
    reAdjustContainerHeight(e?.current);
  };

  const handleImageLoaded = (e) => {
    reAdjustContainerHeight(e?.target);
  };

  const renderContent = () => {
    if (error) {
      return <div>Error! {error.message}</div>;
    }

    if (!imageUrl) {
      return (
        <Skeleton variant="rectangular" className="lazy-image__skeleton" />
      );
    }

    return (
      <Image
        className="lazy-image__img"
        src={imageUrl}
        alt={fileName}
        loading="lazy"
        onLoad={handleImageLoaded}
        onResize={handleImageResized}
      />
    );
  };

  return (
    <div ref={ref} data-inview={inView} style={{ width: "100%", height }}>
      {inView ? renderContent() : null}
    </div>
  );
};

export default LazyImage;
