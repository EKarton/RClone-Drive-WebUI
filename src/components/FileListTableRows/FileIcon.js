import DescriptionIcon from '@mui/icons-material/Description';
import FolderIcon from '@mui/icons-material/Folder';
import ImageIcon from '@mui/icons-material/Image';
import Icon from '@mui/material/Icon';
import PropTypes from 'prop-types';
import LazyImage from 'components/LazyImage';
import { ICON_SIZE } from 'utils/constants';
import './FileIcon.scss';

export default function FileIcon({ file, iconSize, showPreview }) {
  if (file.isDirectory) {
    return <FolderIcon fontSize={iconSize} />;
  }

  if (file.isImage && showPreview) {
    return (
      <Icon fontSize={iconSize} className="file-icon__preview-icon">
        <LazyImage
          image={{
            remote: file.remote,
            dirPath: file.path,
            fileName: file.name,
          }}
          imgClassName="file-icon__img"
        />
      </Icon>
    );
  }

  if (file.isImage) {
    return <ImageIcon fontSize={iconSize} />;
  }

  return <DescriptionIcon fontSize={iconSize} />;
}

FileIcon.propTypes = {
  file: PropTypes.shape({
    isDirectory: PropTypes.bool.isRequired,
    isImage: PropTypes.bool.isRequired,
  }),
  iconSize: PropTypes.oneOf(Object.values(ICON_SIZE)),
  showPreview: PropTypes.bool,
};
