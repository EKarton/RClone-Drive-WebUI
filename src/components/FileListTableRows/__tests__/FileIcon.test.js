import LazyImage from 'components/LazyImage';
import { render } from 'test-utils/react';
import FileIcon from '../FileIcon';

jest.mock('components/LazyImage');

describe('FileIcon', () => {
  it('should match snapshot given file is a directory', () => {
    const file = {
      remote: 'gdrive',
      path: 'Pictures/2021',
      name: 'Screenshots',
      isDirectory: true,
      isImage: false,
    };

    const { baseElement } = render(<FileIcon file={file} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should show an image given file is an image and it is showing previews', () => {
    LazyImage.mockReturnValue(null);

    const file = {
      remote: 'gdrive',
      path: 'Pictures/2021',
      name: 'Big Sur.png',
      isDirectory: false,
      isImage: true,
    };

    const { baseElement } = render(<FileIcon file={file} showPreview />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot given file is an image and it is not showing previews', () => {
    const file = {
      remote: 'gdrive',
      path: 'Pictures/2021',
      name: 'Big Sur.png',
      isDirectory: false,
      isImage: true,
    };

    const { baseElement } = render(<FileIcon file={file} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should match snapshot given file is a pdf', () => {
    const file = {
      remote: 'gdrive',
      path: 'Documents/2021',
      name: 'Forms.png',
      isDirectory: false,
      isImage: false,
    };

    const { baseElement } = render(<FileIcon file={file} />);

    expect(baseElement).toMatchSnapshot();
  });
});
