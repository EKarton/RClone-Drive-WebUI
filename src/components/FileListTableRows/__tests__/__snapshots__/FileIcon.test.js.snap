// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileIcon should match snapshot given file is a directory 1`] = `
<body>
  <div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="FolderIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
      />
    </svg>
  </div>
</body>
`;

exports[`FileIcon should match snapshot given file is a pdf 1`] = `
<body>
  <div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="DescriptionIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
      />
    </svg>
  </div>
</body>
`;

exports[`FileIcon should match snapshot given file is an image and it is not showing previews 1`] = `
<body>
  <div>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="ImageIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M21 19V5c0-1.1-.9-2-2-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2zM8.5 13.5l2.5 3.01L14.5 12l4.5 6H5l3.5-4.5z"
      />
    </svg>
  </div>
</body>
`;

exports[`FileIcon should show an image given file is an image and it is showing previews 1`] = `
<body>
  <div>
    <span
      aria-hidden="true"
      class="material-icons notranslate MuiIcon-root MuiIcon-fontSizeMedium file-icon__preview-icon css-kp9ftd-MuiIcon-root"
    />
  </div>
</body>
`;
